langchain/spark_sql.py

```diff
@@ -0,0 +1,127 @@
+from typing import Any, Iterable, List, Optional
+
+from pyspark.sql import DataFrame, Row, SparkSession
+
+from langchain import SQLDatabase
+
+
+class SparkSQL(SQLDatabase):
+    def __init__(
+        self,
+        spark_session: Optional[SparkSession] = None,
+        catalog: Optional[str] = None,
+        schema: Optional[str] = None,
+        ignore_tables: Optional[List[str]] = None,
+        include_tables: Optional[List[str]] = None,
+        sample_rows_in_table_info: int = 3,
+    ):
+        self._spark = (
+            spark_session if spark_session else SparkSession.builder.getOrCreate()
+        )
+        if catalog is not None:
+            self._spark.catalog.setCurrentCatalog(catalog)
+        if schema is not None:
+            self._spark.catalog.setCurrentDatabase(schema)
+
+        self._all_tables = set(self._get_all_table_names())
+        self._include_tables = set(include_tables) if include_tables else set()
+        if self._include_tables:
+            missing_tables = self._include_tables - self._all_tables
+            if missing_tables:
+                raise ValueError(
+                    f"include_tables {missing_tables} not found in database"
+                )
+        self._ignore_tables = set(ignore_tables) if ignore_tables else set()
+        if self._ignore_tables:
+            missing_tables = self._ignore_tables - self._all_tables
+            if missing_tables:
+                raise ValueError(
+                    f"ignore_tables {missing_tables} not found in database"
+                )
```

Comment:
alternatively we can also use @validator decorator provided by pydantic. :) (link: https://docs.pydantic.dev/latest/usage/validators/)

Reply:
`SQLDatabase` has overlapping logics in `__init__`. Ideally if we migrate to validator, we can re-use many of them, but this would require re-testing of all dependent classes and it will make this PR more complicated. Generally I think it’s good to have each PR serves a single purpose. Please lmk if you want to optimize it, otherwise I’m happy to continue here in a following PR. Thanks! :)

reference: https://github.com/hwchase17/langchain/blob/master/langchain/sql_database.py#L58-L73